### PR TITLE
API docs for the whole stack via doxygen

### PR DESCRIPTION
**Pull request by christophfroehlich**
Instead of running doxygen only on ros2_control repo, I implemented it now for all repositories at once.

This made some changes necessary to the helper scripts: I converted them to python scripts, because it would have been to complicated to formulate the repository-maps with bash scripts.

The doxyfile is moved over from ros2_control, but with updating it to a new doxygen version.

# main page
image

# example controller page
image

### CODE REVIEW

**Comment by christophfroehlich** [comment]
@mergifyio backport humble iron

File: .github/workflows/sphinx-check-page-multiversion.yml [hunk 1]
```diff
@@ -1,7 +1,22 @@
...
   workflow_dispatch:
+    inputs:
+      BASE_BRANCH:
+        description: 'Chose branch to run on'
+        required: true
+        default: 'master'
+        type: choice
+        options:
+        - humble
+        - iron
```
Comment (saikishor) [inline .github/workflows/sphinx-check-page-multiversion.yml:12]: @christophfroehlich should we add jazzy already as a branch option as we know we are going to have it soon?

File: Makefile [hunk 2]
```diff
@@ -30,80 +30,80 @@ html-with-api: Makefile
...
 
 linkcheck-all-subrepos-with-api: Makefile
 	@echo Single version html file with API and linkcheck
 	@echo Step 1: Cloning all subrepositories
-	./make_help_scripts/add_sub_repos
+	./make_help_scripts/add_sub_repos.py
 	@echo Step 2: Building API
-	./make_help_scripts/create_api
-	@echo Step 3: Cloning all subrepositories again
-	./make_help_scripts/add_sub_repos
```
Comment (saikishor) [inline Makefile:71]: Don't we need this cloning subrepositories again script?

File: conf.py [hunk 3]
```diff
@@ -67,10 +67,13 @@
...
 # directories to ignore when looking for source files.
 # This pattern also affects html_static_path and html_extra_path.
-# exclude index.rst files for metapackages for rosdoc2
+# exclude index.rst files from packages/metapackages for rosdoc2
 exclude_patterns = ["_build", "Thumbs.db", ".DS_Store",
                     "**/CHANGELOG.rst", "**/README.rst",
                     "doc/ros2_control/ros2_control/**.rst",
+                    "doc/realtime_tools/doc/**.rst",
+                    "doc/control_msgs/doc/**.rst",
+                    "doc/control_toolbox/doc/**.rst",
```
Comment (saikishor) [inline conf.py:76]: @christophfroehlich What about the package `kinematic_interface`?

File: conf.py [hunk 4]
```diff
@@ -187,9 +190,17 @@
...
+  base_branch = os.environ.get('BASE_BRANCH_PR')
+elif os.environ.get('BASE_BRANCH') is not None:
+  base_branch = os.environ.get('BASE_BRANCH')
+else:
+  base_branch = "master"
 
 # Add branches you want to whitelist here.
-smv_branch_whitelist = r"^(foxy|galactic|humble|iron|master)$"
+smv_branch_whitelist = r"^(foxy|galactic|humble|iron|"+ base_branch + r")$"
 smv_released_pattern = r"^refs/(heads|remotes/[^/]+)/(foxy|galactic|humble|iron).*$"
```
Comment (saikishor) [inline conf.py:204]: Do we still have to whitelist foxy and galactic?

File: index.rst [hunk 5]
```diff
@@ -35,13 +35,13 @@ The ros2_control framework consists of the following Github repositories:
 * `ros2_controllers`_ - widely used controllers, such as forward command controller, joint trajectory controller, differential drive controller;
 * `control_toolbox`_ - some widely-used control theory implementations (e.g. PID) used by controllers;
 * `realtime_tools`_ - general toolkit for realtime support, e.g., realtime buffers and publishers;
-* `control_msgs`_ - common messages.
+* `control_msgs`_ - common messages;
 * `kinematics_interface`_ - for using C++ kinematics frameworks.
```
Comment (saikishor) [inline index.rst:39]:
```suggestion
* `kinematics_interface`_ - for using C++ kinematics frameworks;
```

File: make_help_scripts/deploy_defines.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
+# Copyright (c) 2023 ros2_control maintainers
```
Comment (saikishor) [inline make_help_scripts/deploy_defines.py:1]: @christophfroehlich in this are you checking the commits in other branches? Or only the master branch?

**Review by saikishor — CHANGES_REQUESTED** [review]
Thank you for the changes. They more or less look good to me. I left some minor comments to review.

File: .github/workflows/sphinx-check-page-multiversion.yml (hunk 1 above)
Comment (christophfroehlich) [reply .github/workflows/sphinx-check-page-multiversion.yml:12]: there will be more changes necessary, I'll do this all at once

File: conf.py (hunk 3 above)
Comment (christophfroehlich) [reply conf.py:76]: I haven't added rosdoc2 stuff there yet, but I can add it here.

Comment (christophfroehlich) [inline conf.py:76]:
```suggestion
                    "doc/control_toolbox/doc/**.rst",
                    "doc/kinematics_interface/**.rst",
```

File: conf.py (hunk 4 above)
Comment (christophfroehlich) [reply conf.py:204]: we still published it, but mark them as EOL. I'd leave them for a bit longer..

File: Makefile (hunk 2 above)
Comment (christophfroehlich) [reply Makefile:71]: No, because I removed the cleanup from the `create_api` script

File: make_help_scripts/deploy_defines.py (hunk 6 above)
Comment (christophfroehlich) [reply make_help_scripts/deploy_defines.py:1]: what do you mean, sorry? :grimacing: 

the deploy_defines defines more settings and branch-distro maps for this repo as well as for all subrepos.

File: conf.py (hunk 4 above)
Comment (christophfroehlich) [reply conf.py:205]: I don't change or backport anything to the old distros, just let them be published as they are.

File: Makefile (hunk 2 above)
Comment (saikishor) [reply Makefile:71]: Awesome!

File: conf.py (hunk 4 above)
Comment (saikishor) [reply conf.py:205]: Got it 👍🏽👍🏽👍🏽

File: make_help_scripts/deploy_defines.py (hunk 6 above)
Comment (saikishor) [reply make_help_scripts/deploy_defines.py:1]: Ok I thought this is also doing the PR stats, so I got confused.

**Review by saikishor — APPROVED** [review]
LGTM

**Pull request merged**